src/mococrw/mac.h

```diff
@@ -0,0 +1,138 @@
+ * message authentication codes.
+ */
+class MessageAuthenticationCode {
+public:
+
+    /**
+     * @brief ~MessageAuthenticationCode
+     */
+    virtual ~MessageAuthenticationCode();
+    /**
+     * Adds the message to the hash. This function may be invoked multiple times.
+     * For finishing the calculation of the MAC invoke finish().
+     *
+     * @param message chunk of data used for MAC.
+     * @throws MoCOCrWException if this function is invoked after finish was called.
+     */
+    virtual void update(const std::vector<uint8_t>& message) = 0;
+
+    /**
+     * @brief Finalize the MAC.
+     *
+     * This function calculates the messaged authentication code.
+     *
+     * @throws MoCOCrWException if this function is invoked twice.
+     * @return The calculated message authentication code.
+     */
+    virtual std::vector<uint8_t> finish() = 0;
+
+    /**
+     * @brief Verifies the HMAC
+     *
+     * This function compares the given value (hmacValue), with the value calculated during finish().
+     * Invoke finish() before calling verify().
+     * Only the characters of the shorter value (hmacValue or calculated value) are checked for equality with the
+     * longer one if the length differs.
+     * @throws MoCOCrWException if verify is invoked before finish was invoked.
+     * @throws MoCOCrWException if the verification fails.
+     * @param hmacValue The value which shall be compared to the calculated value
+     */
+    virtual void verify(const std::vector<uint8_t> hmacValue) = 0;
+};
+
+class HMAC : public MessageAuthenticationCode {
+public:
+    /**
+     * @brief Constructor
+     * @param hashFunction The hash function which shall be used
+     * @param key The key used for HMAC
+     * @param outputLength The length of the MAC in bytes
+     */
+    HMAC(mococrw::openssl::DigestTypes hashFunction, const std::vector<uint8_t> &key, size_t outputLength);
+
+    /**
+     * @brief destructor
+     */
+    ~HMAC();
+
+    /**
+     * Adds the chunk of data to the hash. H(i-key || message1 [ || message2])
+     * For getting the HMAC invoke finish()
+     *
+     * @param message chunk of data to used for HMAC.
+     * @throws MoCOCrWException if this function is invoked after finish was called.
+     */
+    void update(const std::vector<uint8_t>& message) override;
+
+
+    /**
+     * @brief Finalize the HMAC.
+     *
+     * This function calculates the message authentication code.
+     * H(o-key || H(i-key || message [ || message]))
+     *
+     * @throws MoCOCrWException if this function is invoked twice.
+     * @return The hashed message authentication code
+     */
+    std::vector<uint8_t> finish() override;
+
+    /**
+     * @see MessageAuthenticationCode::verify
+     */
+    void verify(const std::vector<uint8_t> hmacValue) override;
+
+    /**
+     * @brief The move constructor
+     * @param other The other HMAC to be moved
+     */
+    HMAC(HMAC &&other);
```

Comment:
Should we explicitly delete copying? I don't think that copying such a context with state makes a lot of sense. We should better forbid it in the first place.

Reply:
Done